### PR TITLE
Scalar operators

### LINKED ISSUE #63: Overload comparison operators
Consider overloading comparison operators on `ScalarDistribution` so that `d < 7` create a boolean distribution over True, False. Other sorts of operations like this may be nice to add as well.

### PR DESCRIPTION
**Pull request by Autoplectic**
closes #63 

looks a little like this:

```python
In [1]: dit.ditParams['print.exact'] = dit.ditParams['repr.print'] = True
In [2]: from dit.example_dists import uniform
In [3]: d6 = uniform(1, 7)
In [4]: d6
Out[4]: 
Class:    ScalarDistribution
Alphabet: (1, 2, 3, 4, 5, 6)
Base:     linear

x   p(x)
1   1/6
2   1/6
3   1/6
4   1/6
5   1/6
6   1/6

In [5]: d6 + 3
Out[5]: 
Class:    ScalarDistribution
Alphabet: (4, 5, 6, 7, 8, 9)
Base:     linear

x   p(x)
4   1/6
5   1/6
6   1/6
7   1/6
8   1/6
9   1/6

In [6]: d6 - 1
Out[6]: 
Class:    ScalarDistribution
Alphabet: (0, 1, 2, 3, 4, 5)
Base:     linear

x   p(x)
0   1/6
1   1/6
2   1/6
3   1/6
4   1/6
5   1/6

In [7]: 10 - d6
Out[7]: 
Class:    ScalarDistribution
Alphabet: (4, 5, 6, 7, 8, 9)
Base:     linear

x   p(x)
4   1/6
5   1/6
6   1/6
7   1/6
8   1/6
9   1/6

In [8]: 2*d6
Out[8]: 
Class:    ScalarDistribution
Alphabet: (2, 4, 6, 8, 10, 12)
Base:     linear

x    p(x)
2    1/6
4    1/6
6    1/6
8    1/6
10   1/6
12   1/6

In [9]: d6 % 2
Out[9]: 
Class:    ScalarDistribution
Alphabet: (0, 1)
Base:     linear

x   p(x)
0   1/2
1   1/2

In [10]: (d6 % 2).is_approx_equal(d6 <= 3)
Out[10]: True

In [11]: d6 + d6
Out[11]: 
Class:    ScalarDistribution
Alphabet: (2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12)
Base:     linear

x    p(x)
2    1/36
3    1/18
4    1/12
5    1/9
6    5/36
7    1/6
8    5/36
9    1/9
10   1/12
11   1/18
12   1/36

In [12]: (d6 + d6) % 4
Out[12]: 
Class:    ScalarDistribution
Alphabet: (0, 1, 2, 3)
Base:     linear

x   p(x)
0   1/4
1   2/9
2   1/4
3   5/18

In [13]:  d6 % d6
Out[13]: 
Class:    ScalarDistribution
Alphabet: (0, 1, 2, 3, 4, 5)
Base:     linear

x   p(x)
0   7/18
1   5/18
2   1/6
3   1/12
4   1/18
5   1/36

In [14]: d6 * d6
Out[14]: 
Class:    ScalarDistribution
Alphabet: (1, 2, 3, 4, 5, 6, 8, 9, 10, 12, 15, 16, 18, 20, 24, 25, 30, 36)
Base:     linear

x    p(x)
1    1/36
2    1/18
3    1/18
4    1/12
5    1/18
6    1/9
8    1/18
9    1/36
10   1/18
12   1/9
15   1/18
16   1/36
18   1/18
20   1/18
24   1/18
25   1/36
30   1/18
36   1/36

In [15]: d6 // d6
Out[15]: 
Class:    ScalarDistribution
Alphabet: (0, 1, 2, 3, 4, 5, 6)
Base:     linear

x   p(x)
0   5/12
1   1/3
2   1/9
3   1/18
4   1/36
5   1/36
6   1/36

In [17]:  d6 % (d6 % 2 + 1)
Out[17]: 
Class:    ScalarDistribution
Alphabet: (0, 1)
Base:     linear

x   p(x)
0   3/4
1   1/4
```

### CODE REVIEW

**Comment by Autoplectic** [comment]
i need to add some docstrings, and add a __hash__() method to scalardists.

**Comment by chebee7i** [comment]
Officially awesome!

File: dit/distconst.py [hunk 1]
```diff
@@ -982,3 +982,30 @@ def random_dist_structure(outcome_length, alphabet_size):
...
+    -------
+    d : ScalarDistribution
+        The two distributions combined via `op`
+    """
+    # Copy to make sure we don't lose precision when converting.
+    d2 = d2.copy(base=d1.get_base())
+
+    dist = defaultdict(float)
+    for (o1, p1), (o2, p2) in product(d1.zipped(), d2.zipped()):
+        dist[op(o1, o2)] += d1.ops.mult(p1, p2)
```
Comment (chebee7i) [inline dit/distconst.py]: I was just about to comment that this needs to handle logprobs, but then I remembered...we solved that problem already. Haha.

File: dit/distribution.py [hunk 2]
```diff
@@ -616,6 +626,9 @@ def to_string(self, digits=None, exact=False, tol=1e-9):
         from six import StringIO
         s = StringIO()
 
+        if exact is None:
+            exact = ditParams['print.exact']
```
Comment (chebee7i) [inline dit/distribution.py]: Good idea.

File: dit/divergences/tests/test_generalized_divergences.py [hunk 3]
```diff
@@ -96,11 +96,11 @@ def test_divergences_to_kl():
         for dist1 in dists:
             for dist2 in dists:
                 assert_almost_equal(alpha_divergence(dist1, dist2, alpha=-1), kullback_leibler_divergence(dist2, dist1))
-                if dist1 != dist2:
+                if dist1 is not dist2:
```
Comment (chebee7i) [inline dit/divergences/tests/test_generalized_divergences.py]: Do we want `itertools.pairwise` here?

File: dit/npscalardist.py [hunk 4]
```diff
@@ -489,50 +492,231 @@ def from_distribution(cls, dist, base=None, prng=None, extract=True):
...
-        The appropriate operation is performed assuming that the scalar
-        multiple is of the same base as `self`. If vanilla scalar
-        multiplication is desired, perform the operation directly on
-        `self.pmf`.
+    def __div__(self, other):
+        """
+        Division of ScalarDistribution with either a scalar or another
+        Scalardistribution.
+        """
+        op = lambda x, y: x / y
```
Comment (chebee7i) [inline dit/npscalardist.py]: Just seeking confirmation: The intention here is that this should act like floor division if the user has not imported from `__future__`?

File: dit/divergences/tests/test_generalized_divergences.py (hunk 3 above)
Comment (Autoplectic) [inline dit/divergences/tests/test_generalized_divergences.py]: no, we want something like itertools.product(dists, repeat=2) that skips the collision. pairwise would only return "adjacent" pairs.

Comment (chebee7i) [inline dit/divergences/tests/test_generalized_divergences.py]: Oh right. It should be `combinations` then.

File: dit/npscalardist.py (hunk 4 above)
Comment (Autoplectic) [inline dit/npscalardist.py]: Not exactly, __div__ is called without from __future__ import division on python 2, but __div__ is called for "/" on python 3 always.

note that our test_npscalardist file *does* import division, so on 2.7 only __truediv__ gets tested.

File: dit/divergences/tests/test_generalized_divergences.py (hunk 3 above)
Comment (Autoplectic) [inline dit/divergences/tests/test_generalized_divergences.py]: we do not care about order.

i'll update it to use combinations later. i think there's a spot in MUIProfile that could benefit from this too.

**Comment by Autoplectic** [comment]
it would probably be best if i add an example of op'ing with both a scalar and another scalardist to each docstring.

**Comment by Autoplectic** [comment]
ok, i think we're good to go now.

**Pull request merged**